File: src/match_product_images.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function loadProducts(filePath) {
    const data = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(data);
}

function extractAllSkus(productsData) {
    const allSkus = [];
    
    function traverseCategories(categories) {
        for (const category of categories) {
            if (category.products) {
                for (const product of category.products) {
                    allSkus.push({
                        brandSku: product.brandSku || '',
                        description: product.description || '',
                        specification: product.specification || '',
                        brand: product.brand || '',
                        originalCategory: product.originalCategory || ''
                    });
                }
            }
            if (category.categories) {
                traverseCategories(category.categories);
            }
            if (category.subCategories) {
                for (const subcat of category.subCategories) {
                    if (subcat.categories) {
                        traverseCategories(subcat.categories);
                    }
                }
            }
        }
    }
    
    if (productsData.categories) {
        traverseCategories(productsData.categories);
    }
    
    return allSkus;
}

function matchProductToImage(skuInfo) {
    const sku = String(skuInfo.brandSku || '');
    const description = String(skuInfo.description || '');
    
    const productTypeMapping = {
        'TIP': 'https://via.placeholder.com/400x400/FF6B6B/FFFFFF?text=吸头',
        'PLATE': 'https://via.placeholder.com/400x400/4ECDC4/FFFFFF?text=酶标板',
        'TUBE': 'https://via.placeholder.com/400x400/45B7D1/FFFFFF?text=离心管',
        'BOTTLE': 'https://via.placeholder.com/400x400/96CEB4/FFFFFF?text=试剂瓶',
        'FLASK': 'https://via.placeholder.com/400x400/FFEAA7/333333?text=烧瓶',
        'DISH': 'https://via.placeholder.com/400x400/DDA0DD/FFFFFF?text=培养皿',
        'WELL': 'https://via.placeholder.com/400x400/98D8C8/333333?text=孔板',
        'STRIP': 'https://via.placeholder.com/400x400/F7DC6F/333333?text=酶标条',
        'CUP': 'https://via.placeholder.com/400x400/BB8FCE/FFFFFF?text=离心杯'
    };
    
    for (const [key, productType] of Object.entries(productTypeMapping)) {
        if (sku.toUpperCase().includes(key)) {
            return productType;
        }
    }
    
    if (description.includes('吸头') || description.toLowerCase().includes('tip')) {
        return 'https://via.placeholder.com/400x400/FF6B6B/FFFFFF?text=吸头';
    } else if (description.includes('板') || description.toLowerCase().includes('plate')) {
        return 'https://via.placeholder.com/400x400/4ECDC4/FFFFFF?text=酶标板';
    } else if (description.includes('管') || description.toLowerCase().includes('tube')) {
        return 'https://via.placeholder.com/400x400/45B7D1/FFFFFF?text=离心管';
    } else if (description.includes('瓶') || description.toLowerCase().includes('bottle')) {
        return 'https://via.placeholder.com/400x400/96CEB4/FFFFFF?text=试剂瓶';
    } else if (description.includes('皿') || description.toLowerCase().includes('dish')) {
        return 'https://via.placeholder.com/400x400/DDA0DD/FFFFFF?text=培养皿';
    } else if (description.includes('烧') || description.toLowerCase().includes('flask')) {
        return 'https://via.placeholder.com/400x400/FFEAA7/333333?text=烧瓶';
    }
    
    return 'https://via.placeholder.com/400x400/95A5A6/FFFFFF?text=产品图片';
}

function generateImageMatches(productsFile, outputFile) {
    const productsData = loadProducts(productsFile);
    const allSkus = extractAllSkus(productsData);
    
    const matches = [];
    
    for (const skuInfo of allSkus) {
        const imageUrl = matchProductToImage(skuInfo);
        
        const matchEntry = {
            sku: skuInfo.brandSku,
            description: skuInfo.description,
            imageUrl: imageUrl,
            alternativeImages: []
        };
        
        if (imageUrl && !imageUrl.includes('placeholder')) {
            const baseUrl = imageUrl.substring(0, imageUrl.lastIndexOf('.'));
            for (let i = 2; i <= 4; i++) {
                matchEntry.alternativeImages.push(`${baseUrl}_${i}.jpg`);
            }
        }
        
        matches.push(matchEntry);
    }
    
    fs.writeFileSync(outputFile, JSON.stringify(matches, null, 2), 'utf-8');
    
    console.log(`Generated ${matches.length} image matches`);
    console.log(`Saved to ${outputFile}`);
    
    return matches;
}

const productsFile = path.join(__dirname, 'products_new.json');
const outputFile = path.join(__dirname, 'product_image_matches.json');

try {
    const matches = generateImageMatches(productsFile, outputFile);
    
    console.log('\nSample matches:');
    matches.slice(0, 5).forEach((match, index) => {
        console.log(`\n${index + 1}. SKU: ${match.sku}`);
        console.log(`   Description: ${match.description.substring(0, 50)}...`);
        console.log(`   Image URL: ${match.imageUrl}`);
    });
} catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
}

File: src/match_product_images.py
import json
import re
from typing import Dict, List, Optional

def load_products(file_path: str) -> Dict:
    with open(file_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def extract_all_skus(products_data: Dict) -> List[Dict]:
    all_skus = []
    
    def traverse_categories(categories):
        for category in categories:
            if 'products' in category:
                for product in category['products']:
                    all_skus.append({
                        'brandSku': product.get('brandSku', ''),
                        'description': product.get('description', ''),
                        'specification': product.get('specification', ''),
                        'brand': product.get('brand', ''),
                        'originalCategory': product.get('originalCategory', '')
                    })
            if 'categories' in category:
                traverse_categories(category['categories'])
            if 'subCategories' in category:
                for subcat in category['subCategories']:
                    if 'categories' in subcat:
                        traverse_categories(subcat['categories'])
    
    traverse_categories(products_data.get('categories', []))
    return all_skus

def match_product_to_image(sku_info: Dict) -> Optional[str]:
    sku = sku_info['brandSku']
    description = sku_info['description']
    
    gsbio_base_url = "http://www.gsbio.cn"
    
    product_type_mapping = {
        'TIP': 'pipette-tip',
        'PLATE': 'elisa-plate',
        'TUBE': 'centrifuge-tube',
        'BOTTLE': 'reagent-bottle',
        'FLASK': 'flask',
        'DISH': 'petri-dish',
        'WELL': 'well-plate',
        'STRIP': 'elisa-strip',
        'CUP': 'centrifuge-cup'
    }
    
    for key, product_type in product_type_mapping.items():
        if key in sku.upper():
            return f"{gsbio_base_url}/Uploads/{product_type}.jpg"
    
    if '吸头' in description or 'tip' in description.lower():
        return f"{gsbio_base_url}/Uploads/pipette-tip.jpg"
    elif '板' in description or 'plate' in description.lower():
        return f"{gsbio_base_url}/Uploads/elisa-plate.jpg"
    elif '管' in description or 'tube' in description.lower():
        return f"{gsbio_base_url}/Uploads/centrifuge-tube.jpg"
    elif '瓶' in description or 'bottle' in description.lower():
        return f"{gsbio_base_url}/Uploads/reagent-bottle.jpg"
    elif '皿' in description or 'dish' in description.lower():
        return f"{gsbio_base_url}/Uploads/petri-dish.jpg"
    
    return None

def generate_image_matches(products_file: str, output_file: str):
    products_data = load_products(products_file)
    all_skus = extract_all_skus(products_data)
    
    matches = []
    
    for sku_info in all_skus:
        image_url = match_product_to_image(sku_info)
        
        match_entry = {
            'sku': sku_info['brandSku'],
            'description': sku_info['description'],
            'imageUrl': image_url,
            'alternativeImages': []
        }
        
        if image_url:
            base_url = image_url.rsplit('.', 1)[0]
            for i in range(2, 5):
                match_entry['alternativeImages'].append(f"{base_url}_{i}.jpg")
        
        matches.append(match_entry)
    
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(matches, f, ensure_ascii=False, indent=2)
    
    print(f"Generated {len(matches)} image matches")
    print(f"Saved to {output_file}")
    
    return matches

if __name__ == "__main__":
    products_file = "products_new.json"
    output_file = "product_image_matches.json"
    
    matches = generate_image_matches(products_file, output_file)
    
    print("\nSample matches:")
    for i, match in enumerate(matches[:5]):
        print(f"\n{i+1}. SKU: {match['sku']}")
        print(f"   Description: {match['description'][:50]}...")
        print(f"   Image URL: {match['imageUrl']}")